File: backend/app/schemas/billing.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field, field_validator


class BillingPlanResponse(BaseModel):
    code: str
    name: str
    price_usd: str
    included_minutes: int
    requires_credit_card: bool = True
    limited: bool = False
    sort_order: int = 0
    description: str = ""
    icon: str = ""
    features: list[str] = []
    recommended: bool = False


class BillingPlansListResponse(BaseModel):
    plans: list[BillingPlanResponse]
    billing_provider: str = "manual"


class PaymentMethodInfo(BaseModel):
    brand: str | None = None
    last4: str | None = None
    exp_month: int | None = None
    exp_year: int | None = None


class BillingStatusResponse(BaseModel):
    plan: str | None = None
    status: str | None = None
    minutes_included: int = 0
    minutes_used: int = 0
    minutes_remaining: int = 0
    minutes_carried_over: int = 0
    payment_method: PaymentMethodInfo | None = None
    current_period_end: datetime | None = None
    cancel_at_period_end: bool = False
    has_subscription: bool = False

    model_config = {"from_attributes": True}


class SetupIntentResponse(BaseModel):
    client_secret: str
    customer_id: str


def _validate_plan_code(v: str) -> str:
    from app.services.billing_config_service import get_billing_config

    config = get_billing_config()
    valid = config.valid_plan_codes()
    if v not in valid:
        raise ValueError(f"Plan must be one of: {', '.join(valid)}")
    return v


class SubscribeRequest(BaseModel):
    plan: str
    payment_method_id: str = ""

    @field_validator("plan")
    @classmethod
    def validate_plan(cls, v: str) -> str:
        return _validate_plan_code(v)


class SubscribeResponse(BaseModel):
    plan: str
    status: str
    minutes_included: int
    current_period_end: datetime | None = None


class ChangePlanRequest(BaseModel):
    new_plan: str

    @field_validator("new_plan")
    @classmethod
    def validate_plan(cls, v: str) -> str:
        return _validate_plan_code(v)


class ChangePlanResponse(BaseModel):
    plan: str
    status: str
    minutes_included: int
    minutes_carried_over: int = 0
    current_period_end: datetime | None = None


class CancelResponse(BaseModel):
    status: str
    cancel_at_period_end: bool
    current_period_end: datetime | None = None


class PaymentMethodResponse(BaseModel):
    id: str
    brand: str | None = None
    last4: str | None = None
    exp_month: int | None = None
    exp_year: int | None = None
    is_default: bool = False
    created_at: datetime | None = None

    model_config = {"from_attributes": True}


class PaymentMethodsListResponse(BaseModel):
    payment_methods: list[PaymentMethodResponse]


class AddPaymentMethodRequest(BaseModel):
    payment_method_id: str
    set_as_default: bool = True


class DevSetPlanRequest(BaseModel):
    plan: str
    status: str = "active"

    @field_validator("plan")
    @classmethod
    def validate_plan(cls, v: str) -> str:
        return _validate_plan_code(v)

    @field_validator("status")
    @classmethod
    def validate_status(cls, v: str) -> str:
        valid = ("active", "past_due", "canceled", "incomplete", "trialing")
        if v not in valid:
            raise ValueError(f"Status must be one of: {', '.join(valid)}")
        return v


class DevSimulateUsageRequest(BaseModel):
    minutes: int = Field(..., gt=0)

File: mobile/src/api/billing.ts
import { apiClient } from './client';

export interface BillingPlan {
  code: string;
  name: string;
  price_usd: string;
  included_minutes: number;
  requires_credit_card: boolean;
  limited: boolean;
  sort_order: number;
  description: string;
  icon: string;
}

export interface PaymentMethodInfo {
  brand: string | null;
  last4: string | null;
  exp_month: number | null;
  exp_year: number | null;
}

export interface BillingStatus {
  plan: string | null;
  status: string | null;
  minutes_included: number;
  minutes_used: number;
  minutes_remaining: number;
  minutes_carried_over?: number;
  payment_method: PaymentMethodInfo | null;
  current_period_end: string | null;
  cancel_at_period_end: boolean;
  has_subscription: boolean;
  auto_upgrade_enabled?: boolean;
  auto_upgrade_plan?: string | null;
}

export interface SetupIntentResult {
  client_secret: string;
  customer_id: string;
}

export interface SubscriptionResult {
  plan: string;
  status: string;
  minutes_included: number;
  minutes_carried_over?: number;
  current_period_end: string | null;
}

export interface CancelResult {
  status: string;
  cancel_at_period_end: boolean;
  current_period_end: string | null;
}

export async function getPlans(): Promise<BillingPlan[]> {
  const { data } = await apiClient.get('/billing/plans');
  return data.plans;
}

export async function getBillingStatus(): Promise<BillingStatus> {
  const { data } = await apiClient.get('/billing/status');
  return data;
}

export async function createSetupIntent(): Promise<SetupIntentResult> {
  const { data } = await apiClient.post('/billing/setup-intent');
  return data;
}

export async function subscribe(
  plan: string,
  paymentMethodId: string,
): Promise<SubscriptionResult> {
  const { data } = await apiClient.post('/billing/subscribe', {
    plan,
    payment_method_id: paymentMethodId,
  });
  return data;
}

export async function changePlan(newPlan: string): Promise<SubscriptionResult> {
  const { data } = await apiClient.post('/billing/change-plan', {
    new_plan: newPlan,
  });
  return data;
}

export async function cancelSubscription(): Promise<CancelResult> {
  const { data } = await apiClient.post('/billing/cancel');
  return data;
}

export async function devSetPlan(
  plan: string,
  status: string = 'active',
): Promise<SubscriptionResult> {
  const { data } = await apiClient.post('/dev/billing/set-plan', {
    plan,
    status,
  });
  return data;
}

export async function devSimulateUsage(
  minutes: number,
): Promise<{ minutes_used: number; upgrade_triggered: boolean }> {
  const { data } = await apiClient.post('/dev/billing/simulate-usage-minutes', {
    minutes,
  });
  return data;
}

export interface PaymentMethodItem {
  id: string;
  brand: string | null;
  last4: string | null;
  exp_month: number | null;
  exp_year: number | null;
  is_default: boolean;
}

export async function listPaymentMethods(): Promise<{ items: PaymentMethodItem[] }> {
  const { data } = await apiClient.get('/billing/payment-methods');
  return data;
}

export async function addPaymentMethod(paymentMethodId: string): Promise<PaymentMethodItem> {
  const { data } = await apiClient.post('/billing/payment-methods', { payment_method_id: paymentMethodId });
  return data;
}

export async function removePaymentMethod(methodId: string): Promise<void> {
  await apiClient.delete(`/billing/payment-methods/${methodId}`);
}

export async function setDefaultPaymentMethod(methodId: string): Promise<PaymentMethodItem> {
  const { data } = await apiClient.post(`/billing/payment-methods/${methodId}/default`);
  return data;
}
